File: stops/app/js/optics.js
// Physical optics for the preview and the hand-held floor.
//
// These are the real equations rather than a hand-tuned curve. It matters:
// a preview that shows lush background blur for a 24 mm lens at f/4 teaches
// something false, and the whole point of the app is not to do that.

/** Sensor width in mm, by crop factor relative to 35 mm. */
export function sensorWidth(crop = 1) {
  return 36 / crop;
}

/**
 * Diameter, in mm on the sensor, of the blur disc a background point becomes.
 *
 *   b = (f² / N) · (sb − s) / (s · sb)
 *
 * with focal length f and distances s (subject) and sb (background) in metres,
 * f converted to metres. Returns 0 when the background is at or in front of
 * the subject, and treats an infinite background as the limit sb → ∞.
 */
export function backgroundBlurMm({ focal, aperture, subject, background }) {
  if (!focal || !aperture || !subject) return 0;
  const f = focal / 1000; // mm → m
  if (subject <= f) return 0;
  const ratio = background && isFinite(background)
    ? Math.max(0, (background - subject) / (subject * background))
    : 1 / subject; // background at infinity
  return ((f * f) / aperture) * ratio * 1000; // m → mm
}

/**
 * Length, in mm on the sensor, of the smear a laterally moving subject leaves.
 *
 *   L = (v · t · f) / s
 */
export function motionBlurMm({ focal, shutter, speed, subject }) {
  if (!focal || !shutter || !speed || !subject) return 0;
  const f = focal / 1000;
  return ((speed * shutter * f) / subject) * 1000;
}

/** A length on the sensor, as a fraction of the frame width. */
export function asFrameFraction(mm, crop = 1) {
  return mm / sensorWidth(crop);
}

/**
 * Slowest shutter (in seconds) worth suggesting hand-held.
 *
 * The reciprocal rule gives 1/(focal × crop); stabilisation buys stops on top
 * of it. Whatever the photographer says they personally trust is a hard cap on
 * the result, so the app never quietly suggests something slower than that.
 */
export function handheldFloor({ focal, crop = 1, stabiliserStops = 0, userSlowest = null }) {
  const reciprocal = 1 / (focal * crop);
  const assisted = reciprocal * Math.pow(2, stabiliserStops);
  return userSlowest ? Math.min(assisted, userSlowest) : assisted;
}

/**
 * The shutter speed at which this subject's movement starts to be visible —
 * where its smear crosses a given fraction of the frame width. Useful for
 * telling a beginner what they are actually buying with a faster shutter.
 */
export function motionThreshold({ focal, crop = 1, speed, subject, fraction = 0.006 }) {
  if (!speed || !focal || !subject) return null;
  return (fraction * sensorWidth(crop) * subject) / (speed * focal);
}

/** The 500 rule: longest exposure before stars trail, in seconds. */
export function starTrailLimit({ focal, crop = 1 }) {
  return 500 / (focal * crop);
}

/**
 * Widest aperture a zoom offers at a given focal length. Real zooms step, but
 * interpolating in stops across the range is close enough to be useful and
 * never optimistic by more than a third of a stop on common kit lenses.
 */
export function widestAt(lens, focal) {
  if (!lens) return 2.8;
  const { min, max, wideMin, wideMax } = lens;
  if (wideMax == null || wideMin === wideMax || min === max) return wideMin;
  const clamped = Math.min(Math.max(focal, min), max);
  const position = Math.log2(clamped / min) / Math.log2(max / min);
  const stops = Math.log2(wideMax / wideMin) * position;
  return wideMin * Math.pow(2, stops);
}

File: stops/ios/OffAutoKit/Sources/OffAutoKit/Optics.swift
import Foundation

/// The physical optics the guide's two computed tables rest on.
///
/// The real equations rather than a hand-tuned curve, for the same reason the
/// web app uses them: a table that says a 24 mm lens needs f/8 to hold two
/// people sharp when it does not is teaching something false, and not doing
/// that is the whole point.
public enum Optics {
  /// Sensor width in mm. Full frame, since that is the camera every scene is
  /// written for and a crop factor is one more thing to explain on day one.
  public static let sensorWidth: Double = 36
  public static var circleOfConfusion: Double { sensorWidth / 1500 }

  /// Slowest shutter worth suggesting hand-held: the reciprocal rule, with
  /// stabilisation buying stops on top of it.
  public static func handheldFloor(focal: Double, stabiliserStops: Double = 0) -> Double {
    (1 / focal) * pow(2, stabiliserStops)
  }

  /// The shutter at which this subject's movement starts to show, where its
  /// smear crosses a given fraction of the frame width.
  public static func motionThreshold(focal: Double, speed: Double, subject: Double,
                                     fraction: Double = 0.006) -> Double? {
    guard speed > 0, focal > 0, subject > 0 else { return nil }
    return (fraction * sensorWidth * subject) / (speed * focal)
  }

  /// The f-number needed to hold something at `far` inside the circle of
  /// confusion while focused at `subject`.
  public static func apertureForDepth(focal: Double, subject: Double, far: Double) -> Double? {
    guard focal > 0, subject > 0, far > subject else { return nil }
    let f = focal / 1000
    let ratio = far.isFinite ? (far - subject) / (subject * far) : 1 / subject
    return ((f * f) * ratio * 1000) / circleOfConfusion
  }
}

/// Things that move, and how fast, for the shutter page.
public struct Mover: Hashable, Sendable {
  public let name: String
  public let speed: Double
  /// How far away it sits when framed at 50 mm. Distance scales with focal
  /// length to hold the framing, and the focal length then cancels out of the
  /// sum exactly, which is the surprising part worth teaching.
  public let at50: Double
}

public enum Movers {
  public static let all: [Mover] = [
    Mover(name: "Someone posing", speed: 0.3, at50: 2),
    Mover(name: "Someone walking", speed: 1.4, at50: 5),
    Mover(name: "A child or a dog", speed: 3, at50: 3),
    Mover(name: "A runner", speed: 6, at50: 8),
    Mover(name: "A cyclist going past", speed: 8, at50: 10),
    Mover(name: "A bird in flight", speed: 12, at50: 20),
  ]

  public static func threshold(focal: Double, speed: Double, subject: Double) -> Double {
    Optics.motionThreshold(focal: focal, speed: speed, subject: subject) ?? 1
  }
}

/// Groups of things that all have to be sharp, for the aperture page.
public struct Depth: Hashable, Sendable {
  public let name: String
  public let at50: Double
  /// The gap between the near and far thing that both have to be sharp. It does
  /// not scale with focal length: two rows of people are a metre apart whatever
  /// lens is pointed at them.
  public let gap: Double
}

public enum Depths {
  public static let all: [Depth] = [
    Depth(name: "Both eyes on one face", at50: 1.5, gap: 0.08),
    Depth(name: "Two people side by side", at50: 2.5, gap: 0.4),
    Depth(name: "Two rows of people", at50: 4, gap: 1),
    Depth(name: "A table of six", at50: 2.5, gap: 1.6),
  ]

  public static func aperture(focal: Double, subject: Double, far: Double) -> Double? {
    Optics.apertureForDepth(focal: focal, subject: subject, far: far)
  }
}
